components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java

```diff
@@ -75,53 +75,45 @@ public static AuthorizationHandlerManager getInstance() throws IdentityOAuth2Exc
+	    // IDENTITY-4111
+	    OAuth2Util.setAuthzRequestContext(authzReqMsgCtx);
+	    authorizeRespDTO = authzHandler.issue(authzReqMsgCtx);
+        } finally {
+            // clears authorization request context
+            OAuth2Util.clearAuthzRequestContext();
+        }
+        return authorizeRespDTO;
+    }
 
-        ResponseTypeHandler authzHandler = responseHandlers.get(responseType);
-        OAuthAuthzReqMessageContext authzReqMsgCtx = new OAuthAuthzReqMessageContext(authzReqDTO);
-
-        // loading the stored application data
-        OAuthAppDO oAuthAppDO = getAppInformation(authzReqDTO);
-        authzReqMsgCtx.addProperty("OAuthAppDO", oAuthAppDO);
-
-        // load the SP tenant domain from the OAuth App info
-        authzReqMsgCtx.getAuthorizationReqDTO().setTenantDomain(OAuth2Util.getTenantDomainOfOauthApp(oAuthAppDO));
-
-        boolean isAuthorizedClient = authzHandler.isAuthorizedClient(authzReqMsgCtx);
-
-        if (!isAuthorizedClient) {
-            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNAUTHORIZED_CLIENT,
-                    "The authenticated client is not authorized to use this authorization grant type");
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+    private OAuth2AuthorizeRespDTO validateAuthzRequest(OAuth2AuthorizeReqDTO authzReqDTO,
+                                                        OAuthAuthzReqMessageContext authzReqMsgCtx,
+                                                        ResponseTypeHandler authzHandler)
+            throws IdentityOAuth2Exception {
+        OAuth2AuthorizeRespDTO authorizeRespDTO = new OAuth2AuthorizeRespDTO();
+        if(!validateResponseType(authzReqDTO, authorizeRespDTO)) {
             return authorizeRespDTO;
         }
-
-        boolean accessDelegationAuthzStatus = authzHandler.validateAccessDelegation(authzReqMsgCtx);
-        if(authzReqMsgCtx.getProperty("ErrorCode") != null){
-            authorizeRespDTO.setErrorCode((String)authzReqMsgCtx.getProperty("ErrorCode"));
-            authorizeRespDTO.setErrorMsg((String)authzReqMsgCtx.getProperty("ErrorMsg"));
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+        if (!validateClient(authzReqDTO, authorizeRespDTO, authzReqMsgCtx, authzHandler)) {
             return authorizeRespDTO;
-        } else if (!accessDelegationAuthzStatus) {
-            log.warn("User : " + authzReqDTO.getUser() +
-                    " doesn't have necessary rights to grant access to the resource(s) " +
-                    OAuth2Util.buildScopeString(authzReqDTO.getScopes()));
-            handleErrorRequest(authorizeRespDTO, OAuthError.CodeResponse.UNAUTHORIZED_CLIENT,
-                    "Authorization Failure!");
-            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+        }
+        if (!validateAccessDelegation(authzReqDTO, authorizeRespDTO, authzReqMsgCtx, authzHandler)) {
```

Comment:
Better to have a debug log

Reply:
Added debug logs within validateAccessDelegation() method